agent.go

```diff
@@ -622,44 +623,116 @@ type initMount struct {
+		{"bind", "/sys/fs/cgroup/cpu,cpuacct", "/sys/fs/cgroup/cpu", []string{"bind"}},
+		{"bind", "/sys/fs/cgroup/cpu,cpuacct", "/sys/fs/cgroup/cpuacct", []string{"bind"}},
+		{"bind", "/sys/fs/cgroup/net_cls,net_prio", "/sys/fs/cgroup/net_cls", []string{"bind"}},
+		{"bind", "/sys/fs/cgroup/net_cls,net_prio", "/sys/fs/cgroup/net_prio", []string{"bind"}},
+		{"tmpfs", "tmpfs", "/sys/fs/cgroup", []string{"remount", "ro", "nosuid", "nodev", "noexec", "mode=755"}},
+	}
+
+	procCgroups = "/proc/cgroups"
+	cgroupPath  = "/sys/fs/cgroup"
+)
+
+func getCgroupMounts(cgPath string) ([]initMount, error) {
+	f, err := os.Open(cgPath)
+	if err != nil {
+		return []initMount{}, err
+	}
+	defer f.Close()
+
+	cgroupMounts := []initMount{{"tmpfs", "tmpfs", cgroupPath, []string{"nosuid", "nodev", "noexec", "mode=755"}}}
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		text := scanner.Text()
+		fields := strings.Split(text, "\t")
+		cgroup := fields[0]
+
+		if cgroup == "" || cgroup[0] == '#' || (len(fields) > 3 && fields[3] == "0") {
+			continue
+		}
+		cgroupMounts = append(cgroupMounts, initMount{"cgroup", "cgroup",
+			filepath.Join(cgroupPath, cgroup), []string{"nosuid", "nodev", "noexec", "relatime", cgroup}})
+	}
+
+	if err = scanner.Err(); err != nil {
+		return []initMount{}, err
+	}
+
+	cgroupMounts = append(cgroupMounts, initMount{"tmpfs", "tmpfs",
+		cgroupPath, []string{"remount", "ro", "nosuid", "nodev", "noexec", "mode=755"}})
+	return cgroupMounts, nil
 }
 
-// initAgentAsInit will do the initializations such as setting up the rootfs
-// when this agent has been run as the init process.
-func initAgentAsInit() error {
+func mountToRootfs(m initMount) error {
+	if err := os.MkdirAll(m.dest, os.FileMode(0755)); err != nil {
+		return err
+	}
+
+	if flags, options, err := parseMountFlagsAndOptions(m.options); err != nil {
+		return grpcStatus.Errorf(codes.Internal, "Could parseMountFlagsAndOptions(%v)", m.options)
+	} else if err := syscall.Mount(m.src, m.dest, m.fstype, uintptr(flags), options); err != nil {
+		return grpcStatus.Errorf(codes.Internal, "Could not mount %v to %v: %v", m.src, m.dest, err)
+	}
+	return nil
+}
+
+func generalMount() error {
 	for _, m := range initRootfsMounts {
 		if err := os.MkdirAll(m.dest, os.FileMode(0755)); err != nil {
 			return err
 		}
-		if flags, options, err := parseMountFlagsAndOptions(m.options); err != nil {
-			return grpcStatus.Errorf(codes.Internal, "Could parseMountFlagsAndOptions(%v)", m.options)
-		} else if err = syscall.Mount(m.src, m.dest, m.fstype, uintptr(flags), options); err != nil {
-			return grpcStatus.Errorf(codes.Internal, "Could not mount %v to %v: %v", m.src, m.dest, err)
+		if err := mountToRootfs(m); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func cgroupsMount() error {
+	cgroups, err := getCgroupMounts(procCgroups)
+	if err != nil {
+		cgroups = initCgroupMounts
```

Comment:
I don't think we need to fallback here. If getCgroupMounts fails, there is no cgroups support in the kernel at all and kata agent should fail to start. And the fallback will fail as well. So there is no need for it.

Reply:
Whether getCgroupMounts  is successful or not depend on mount /proc before getCgroupMounts ,if agent getCgroupMounts  before mount /proc,  getCgroupMounts failed， but the kernel still support cgroup.

in this pr getCgroupMounts comes after mounting /proc, so we need not fallback，need me remove it ?